### ISSUE TITLE
RecursionError in pluggy

### ISSUE THREAD

**Issue by The-Compiler** [pytest-dev/pytest#1794]
When I try to run the [qutebrowser](https://github.com/The-Compiler/qutebrowser) testsuite with the current git `features` branch, I get:

```
_____ ERROR at teardown of test_timeout_set_interval _____

hook = <_HookCaller 'pytest_fixture_post_finalizer'>
methods = [<_pytest.vendored_packages.pluggy.HookImpl object at 0x7f3dd6aa7b38>]
kwargs = {'fixturedef': <FixtureDef name='httpbin' scope='session' baseid='tests/end2end' >}

    self._inner_hookexec = lambda hook, methods, kwargs: \
>       _MultiCall(methods, kwargs, hook.spec_opts).execute()
E   RecursionError: maximum recursion depth exceeded

.../_pytest/vendored_packages/pluggy.py:333: RecursionError
```

`test_timeout_set_interval` seems unrelated, as this is just the teardown of a session-scoped fixture. However, when I only run pytest on `tests/end2end` I can't reproduce this...

The `httpbin` fixture is [defined here](https://github.com/The-Compiler/qutebrowser/blob/master/tests/end2end/fixtures/webserver.py#L169-L175).

I have no idea what's happening though... Taking the freedom to mark this as critical as it might be a serious 3.0 regression for other testsuites too.

**Comment by nicoddemus** [comment]
@The-Compiler, thanks, I plan to start investigating this tomorrow. It seems related to the new `pytest_fixture_post_finalizer` hook.

Meanwhile, if you have the time, could you try to bisect this from 2.9.2 and try to find the culprit commit?

**Comment by The-Compiler** [comment]
Trying to bisect, but I ran into some trouble with this not being reproducible anymore first (not sure why). I think things are looking better now, but it'll still take me a bit.

**Comment by The-Compiler** [comment]
Bisected to 775100881a547fe1a03299a3e4754662dd02b9fc ("Implement invocation-scoped fixtures"). @nicoddemus, your turn :wink:

**Comment by nicoddemus** [comment]
Thanks for the bisect! :grin: 

I will try to tackle this tonight.

Is there a simple non-obvious command to reproduce the error in `qutebrowser`, for example `py.test tests/some_test_file.py --someoption`, or does it happen only when running the full test suite?

**Comment by The-Compiler** [comment]
Here's roughly what I did:

- create/activate a virtualenv
- `pip install git+https://github.com/pytest-dev/pytest.git@features`
- `pip install -rrequirements.txt`
- `/usr/bin/python3 scripts/link_pyqt.py .venv`
- edit `conftest.py` to always skip BDD tests in `pytest_ignore_collect` (simply set `skip_bdd = True`)
- `py.test`

I initially tried running a subset but then the error went away, so I guess better save than sorry. Still takes around 1min 40s per run though...

**Comment by nicoddemus** [comment]
On this for some hours now, made some progress but still no idea what's the problem. :sob: I'll just dump some information here in the hope that sparks some *heureka* moment in someone.

I could not reproduce the problem on my Windows computer because of `PyQt5.QtWebKit` not being easily available and not being in the mood to compile it myself, so I had to investigate this on a spare Unix box which didn't have my usual dev environment on it, so progress has been slow. :pensive:

Anyways, I've tried to reduce the number of tests one had to run to find the issue, and I managed to remove quite a few test files until I had a small enough set of tests that still manifested the problem, reducing the test suite time to 8 seconds. I [pushed a branch](https://github.com/nicoddemus/qutebrowser/tree/pytest-issue-1794) to my fork in order to continue this tomorrow or if others want to take a look.

Some facts:

* When commenting out the `pytest_fixture_post_finalizer` hook call in `fixtures.py`, the traceback changes and shows where the actual recursion is happening:

  ```
self = <FixtureDef name='qapp' scope='session' baseid='tests' >                                                                                                                                                                           
                                                                                                                                                                                                                                          
    def finish(self):                                                                                                                                                                                                                     
        try:                                                                                                                                                                                                                              
            while self._finalizer:                                                                                                                                                                                                        
                func = self._finalizer.pop()                                                                                                                                                                                              
>               func()                                                                                                                                                                                                                    
E               RuntimeError: maximum recursion depth exceeded                                                                                                                                                                            
                                                                                                                                                                                                                                          
../pytest/_pytest/fixtures.py:751: RuntimeError            
  ```

  So it seems some finalizer is registering itself, *somehow*.

* It's definitely related to either or both `httpbin` and `httpbin_after_test` fixtures. On my branch if I remove their `autouse` status or don't execute `tests/end2end/test_dummy.py` (which exists solely that `tests/end2end/conftest.py` is considered) the error no longer happens.

* `httpbin` and other fixtures were being *imported* directly into `tests/end2end/conftest.py`:

 ```python
from end2end.fixtures.webserver import httpbin, httpbin_after_test, ssl_server
  ```

  This might be a problem because depending on where they are used they might be initialized/finalized more than once because pytest will think they are defined in different places and are different fixtures. The correct approach would be to use `pytest_plugins` variable to declare those, but that doesn't seem to be the issue here: I did try to rename the fixture function to `httpbin_` and mark it with `@pytest.fixture(name='httpbin', ...` to see if other places imported it, but doesn't seem to be the case.

* I couldn't really manage to see which actual function was the `func` variable in the error traceback; I focused more on trying to narrow the suite, but I guess that is a good starting point for tomorrow.

Well that's what I can remember now with my tired brain at this late hour. I plan to continue debugging this tomorrow, but any comments or insights are welcome. :sweat_smile:

**Comment by The-Compiler** [comment]
I can confirm that subset reproduces it for me. I could also get rid of `test_split_hypothesis.py` and `test_log.py`.

When I comment out `ihook.pytest_fixture_post_finalizer(fixturedef=self)` then, I don't get a `RecursionError` anymore. After re-adding the test files I removed above, it's there again.

So it sounds like it's - pure speculation - something with the stack getting bigger with more test items, and then at some point too big?

**Comment by The-Compiler** [comment]
FYI: I edited the post above multiple times to provide more and more information - so don't view it via email  :wink:

**Comment by nicoddemus** [comment]
@The-Compiler thanks a lot for the debugging help! I will continue this tonight. :sweat_smile: 

Me this morning: *"Hmm if only I knew of a project which used PyQt5 and QtWebKit on Windows, so I can see which package they used to install it hassle-free..."*... duh. :laughing:

**Comment by The-Compiler** [comment]
Installing Python 3.4 and PyQt 5.5.1 (e.g. from [here](https://www.qutebrowser.org/pyqt/), or probably still available on sourceforge) might be the easiest way.

**Comment by nicoddemus** [comment]
Yep, the *face-palm* moment was due to the fact that I could just have seen how [qutebrowser did it](https://github.com/The-Compiler/qutebrowser/blob/master/scripts/dev/ci/appveyor_install.py#L50). :grin:

**Comment by The-Compiler** [comment]
Right - seems like we both need more :coffee: :wink:

**Comment by nicoddemus** [comment]
Quick update: I've made some progress, but I'm still in the process of reproducing this in an isolated test. :grimacing:

**Comment by nicoddemus** [comment]
(@The-Compiler I changed the title of this issue, I hope that's OK)

**Comment by nicoddemus** [comment]
When this is tackled, make sure to revert the fix for #1872

**Comment by blueyed** [comment]
What the state of this?
Any progress on the front of this nice feature?
Is there a PR / branch where it's being worked on?

**Comment by nicoddemus** [comment]
> Any progress on the front of this nice feature?

No progress, unfortunately. 

There were two approaches that were discussed/implemented:

1. Change the fixture caching mechanism to also cache based on scope; this was tried during the sprint but hit some roadblocks.
2. Group `FixtureDef` on the `Request` objects not only by name, but also by scope. That was the implementation I had for 3.0 but broke down when faced with fixtures being overwritten in the same name.

I gave it another try right after 3.0 was out but I'm not sure it is possible with the current architecture. We would probably have to rethink how things are implemented in order to support this in all the possible use cases.

In summary, I don't think this will land in 3.1 unless someone else gives it a shot. 😞

**Comment by blueyed** [comment]
Too bad, but don't feel 😞! :)

Is there some WIP branch somewhere?

I've came back to this because I've started for at least the 2nd time to have some `foo_session` fixtures etc - basically what this would automate.

**Comment by RonnyPfannschmidt** [comment]
@blueyed main reason there is no wip is, that the current internals of fixtures make multi scoped fixtures on top of them a mess that can easily break apart

my current suspicion is that we need to rewrite the internals from ground up to enable it :(

**Comment by nicoddemus** [comment]
@blueyed for the record I pulled it back in a single commit: 707b6b5e3fee6626c9ea94 (well and e92d373460397caeb but that was just docs), it should be possible to revert it and continue from there.

**Comment by nicoddemus** [comment]
We have reverted invocation scoped fixtures and I've reopened the original issue in #1681, so I'm closing this for now.
